### ISSUE TITLE
Data contracts

### ISSUE THREAD

**Issue by ng-druid** [rollthecloudinc/hedge#103]
A contract is an obligation for data to adhere to certain requirements. For example, the object is required to have a title that is a string no longer than 125 characters. Contracts are how hedge enforces data integrity for any document being stored. Without contracts anyone can upload anything. At the moment this isn't a huge concern since uploading is only available to github app repo owners and those that are granted explicit permission to persist via repo owners. However, the intent is to open this system up to uploading data via anonymous users or rather those that are just using a website. Without some way to enforce data integrity anyone could upload anything. What we want instead is to maintain consistency between objects uploaded at different paths. For example, an object that is an ad should have all the requirements of an ad enforced by a contract. Contracts can also include permission info. Provide grants for specific users, etc. to only upload even read data.

Enforce data contracts at the path level. Data contracts should also be able to cover everything underneath a path when desired. The most straightforward contract will use json schema to validate incoming json. Data contracts can also be variable based on rules. For example, different contracts based on a single field value like type.

func/enforce_contract
func/matches_schema
func/validate_schema

The function will receive the entity and json schema verifying that the data satisfies the json schema requirements.

The library that will be used for validating the provided json schema against entity.

https://ajv.js.org/guide/getting-started.html

I like the idea of using stand alone validation. When a user uploads a schema file process kicks off to generate stand alone validator to make validation much more performant.

https://ajv.js.org/standalone.html

JSON rules engine rules can be used to predetermine the schema to apply to incoming data.

https://www.npmjs.com/package/json-rules-engine

* schema.json – schema file
* contract.json – configuration for data contract

Example:

* contract.json
* ad.schema.json
* realestate.schema.json
* auto.schema.json
* apartment.schema.json

Contact config will have location of files inside. The schema files can be anywhere in the repo. Perhaps even remotely located. Remote schemas would make it possible to do some very dynamic things.

There should also be an api end-point for testing data contract against data without writing. 

Data will also be flagged with the version of the contract the data was validated against. The entity will also have some type of ref to the original contract perhaps even embed the contract in the entity.

* user uploads schema *.schema.json
* Lambda compiles to new file excluding ajm 
* *.schema.js
* Entity uploaded
* Validation uses schema
* 

Contracts can be uploaded by owners but require explicit permissions for everyone else with write access. Contracts can never be uploaded by unauthorized users even though they will eventually be able to upload entities.

For now heavily leaning towards github actions.

https://github.com/marketplace/actions/add-commit

We would add a workflow file to every spearhead-objects and every spearhead-objects-prod repo that uses the ajm cli to generate a file with the validation functions from the uploaded schema.

The validation lambda is written in js accepts the path to the schema and array of entities. Returns validation result object to each entity. The validation lambda with loop through the entity array and execute the precompiled validation functions send the results back to the process that initiated the lambda for each entity in an array.

Execute js using go

https://github.com/rogchap/v8go

Github action to get all changed files.

https://github.com/marketplace/actions/get-all-changed-files

----------------

Contract Discovery

The contract discovery system should be very flexible but we can start off with a simple implementation. In fact the contract discovery system might even be best suited to a new entity adaptor.

Simple discovery for now.

either folder at root as:

* entities
* contracts

Each contract will contain the path which the contract applies:

ie. 

* path: /ads/* = one level under ads
* path: /ads/** = anything under ads hierarchy

We may want to consider storing contracts also in cassandra to make it easy to query for them. Otherwise we will need to create a function to query all contracts under the directory using github restful api which can be slow. Hitting cassandra will be much faster than querying github. To that end we could also consider storing all data in cassandra… if we do that than we can also implement search without needing opensearch which would be the cheaper option.

-----------------

## Strategy 1

Relies on Github actions and ajv cli to generate standalone JavaScript validation function for schema. Requires adding workflow file to every objects repository. The workflow will detect files ending with .schema.json run through ajv cli generating standalone JavaScript validation file. The validation file will be committed back to the repository in the action. Both the schema and JavaScript validation become part of repository with all the advantages of version control.

HEDGE will use either an auto detection strategy and/or contract.json to discovery schema consequently stand alone validation JavaScript files. The javascript file will be executed using golang to validate incoming objects.

The primary disadvantage of this strategy is the tight coupling to Github. Also the requirement of adding the lib to run JavaScript using golang. Although the size/impact of adding that lib isn't known at this time.

For now me might just implement discovery in the path as an expected file name like entity.schema.json.

* entity.schema.json
* entity.schema.js

------------

## Strategy 2

Create js lambda with object and schema as input that returns validation result. The validation adaptor will call this function from go. The one caveat with this approach is also making is compatible with azure. Fore azure a new serverless config will be required and probably some modifications to make it compatible. We also need to figure out how to call this new azure function app from the primary hedge azure function app.

------------

Example Contract:

<img width="1440" alt="Screen Shot 2023-05-03 at 5 19 10 AM" src="https://user-images.githubusercontent.com/73197190/235878162-12941167-1742-4dec-9334-09170f751282.png">

<img width="1261" alt="Screen Shot 2023-05-03 at 4 41 10 AM" src="https://user-images.githubusercontent.com/73197190/235877979-90fc2652-fb6b-451c-afd3-fb5c6c3f1fff.png">

https://github.com/rollthecloudinc/hedge-objects/blob/dev/data-contracts/person.schema.json

----------------------------------------------

Next Items:

* Azure compatibility
  * Legacy Validation (golang)
    * struct json validation
   * contract enforcement (javascript)
     * json schema validation
       * Fortunately serverless framework and azure support JavaScript runtime in azure.
* Validate schema's option for enforce_contact - identify as schema that needs to be validated
* Carbon metrics logging – enforce_contract lambda
  * AWS
  * Azure 
    * Separate topic since no logging exists yet for this anywhere on azure. Requires R&D
* [optional] Read, write, delete, authorization rules and enforcement. Perhaps this belongs as separate issue…

--------------------

Azure Compatibility

After conducting research the approach to use will be providing a dedicated rest api as a separate function that validates the schema.

https://www.serverless.com/framework/docs/providers/azure/events/http

**Comment by ng-druid** [comment]
This is going to be required for Canva. We will probably need to cherry pick the merge request once 155 is merged. 155 should be in a good state ready to merge before tackling integrating this back in. The current mr works fine for the older environment but not for the new al2023 environment we have been forced to migrate to.

**Comment by ng-druid** [comment]
The MVP for contracts has been merged. The MVP is limiting ti supporting contracts with static schemas. The system expects contracts to exist in the contracts directory in the root of a repo. Any contracts existing outside this location will be ignored. Furthermore, the contract system is limited to using the first path part after shapeshifter to match the request entity to a contract.
